fuel-types/src/canonical.rs

```diff
@@ -0,0 +1,619 @@
+    fn size(&self) -> usize {
+        0
+    }
+
+    #[inline(always)]
+    fn encode_static<O: Output + ?Sized>(&self, _buffer: &mut O) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+impl Deserialize for () {
+    fn decode_static<I: Input + ?Sized>(_buffer: &mut I) -> Result<Self, Error> {
+        Ok(())
+    }
+}
+
+/// To protect against malicious large inputs, vectors size is limited on decoding.
+pub const VEC_DECODE_LIMIT: usize = 100 * (1 << 20); // 100 MiB
+
+impl<T: Serialize> SerializedSize for Vec<T> {
+    const SIZE_STATIC: usize = 8;
+}
+
+impl<T: Serialize> Serialize for Vec<T> {
+    const SIZE_NO_DYNAMIC: bool = false;
+
+    #[inline(always)]
+    // Encode only the size of the vector. Elements will be encoded in the
+    // `encode_dynamic` method.
+    fn encode_static<O: Output + ?Sized>(&self, buffer: &mut O) -> Result<(), Error> {
+        assert!(
+            self.len() < VEC_DECODE_LIMIT,
+            "Refusing to encode vector too large to be decoded"
+        );
+        let len: u64 = self.len().try_into().expect("msg.len() > u64::MAX");
+        len.encode(buffer)
+    }
+
+    fn encode_dynamic<O: Output + ?Sized>(&self, buffer: &mut O) -> Result<(), Error> {
+        // Bytes - Vec<u8> it a separate case without padding for each element.
+        // It should padded at the end if is not % ALIGN
+        if T::UNALIGNED_BYTES {
+            // SAFETY: `UNALIGNED_BYTES` only set for `u8`.
+            let bytes = unsafe { ::core::mem::transmute::<&Vec<T>, &Vec<u8>>(self) };
+            buffer.write(bytes.as_slice())?;
+            for _ in 0..alignment_bytes(self.len()) {
+                buffer.push_byte(0)?;
+            }
+        } else {
+            for e in self.iter() {
+                e.encode(buffer)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<T: Deserialize> Deserialize for Vec<T> {
+    // Decode only the capacity of the vector. Elements will be decoded in the
+    // `decode_dynamic` method. The capacity is needed for iteration there.
+    fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
+        let cap = u64::decode(buffer)?;
+        let cap: usize = cap.try_into().map_err(|_| Error::AllocationLimit)?;
+        if cap > VEC_DECODE_LIMIT {
+            return Err(Error::AllocationLimit)
+        }
+        Ok(Vec::with_capacity(cap))
+    }
+
+    fn decode_dynamic<I: Input + ?Sized>(&mut self, buffer: &mut I) -> Result<(), Error> {
+        for _ in 0..self.capacity() {
+            // Bytes - Vec<u8> it a separate case without unpadding for each element.
+            // It should unpadded at the end if is not % ALIGN
+            if T::UNALIGNED_BYTES {
+                let byte = buffer.read_byte()?;
+                // SAFETY: `UNALIGNED_BYTES` implemented set for `u8`.
+                let _self =
+                    unsafe { ::core::mem::transmute::<&mut Vec<T>, &mut Vec<u8>>(self) };
+                _self.push(byte);
```

Comment:
why do we take the first byte and append it at the end?

Reply:
We iterate over all bytes(we use `capacity`, not length to do that because we have an empty vector with a pre-defined capacity) and add them to the current vector. Everything is expected=)

Reply:
Yep, this is indeed correct